### ISSUE TITLE
wxGrid doesn't seem to work with screen readers used by visually impaired programmers (JAWS or NVDA)

### ISSUE THREAD

**Issue by jcprovost** [wxGlade/wxGlade#534]
Hi,

the wxGrid control seem to be inaccessible when I tried to use it both in my program & in the %wxGlade 1.0.5%\examples\AllWidgets\AllWigets_30.wxg [preview] mode.

Messing around with the properties of the control did not help.

Behavior: after I locate the wxGrid control I press <tab> to get to it, then I can't do anything anymore.  Neither the [tab] nor the [arrow] keys give me any feedback.  I can't even get back to the notebook pages.  The only way out is to exit the [preview] mode.

Expected behavior: I should be able to move inside the grid cells with the [arrow] keys, as well as with [home], [end], [ctrl+home], [ctrl+end] & even [pgup] & [pgdn].

Will gladly provide more assistance/clarifications, should any be needed to resolve this issue.

Failing that, I'll entertain any work around you may offer.

Thanks in advance for your assistance.

Cheers,
@DietmarSchwertberger

**Comment by DietmarSchwertberger** [comment]
In the last days I had a look at the wxWidgets sources. I think I will add accessibility support for wxGrid.
After implementation, it will still take some time until this finds its way into wxPython.

**Comment by jcprovost** [comment]
YEAH!!! 😊

Now that, is wonderful news!  Thanks a bunch for taking the time.  I make myself a spokesperson for the Visually impaired programmers community to send a big & warm appreciative wave down your neck of the woods😊

Because I’m not very familiar with the implementation cycle, I’m assuming we won’t be able to use your improvement until XwPython implements it at their end?  I just saw the issue getting closed so, I’m assuming your part is completed.  Is their an easy way for me to know when WxPython has completed theirs?

Thanks again,

Jean-Claude Provost

**Comment by DietmarSchwertberger** [comment]
I have closed the ticket here as there is no work on wxGlade involved.

At the moment I'm learning by looking at the wxDataViewCtrl implementation, which supports accessibilty already.
Most wxWidget controls have accessibility support out of the box because they use native Windows controls.
wxGrid and wxDataViewCtrls are generic widgets, rendered by wxWidgets itself and therefore don't inherit the native support from Windows.

I think I will find the time to implement the wxWidgets changes at latest over Christmas vacation. 

Once I have the wxWidgets modification, I plan to upload somewhere a version of the wxWidgets grid sample as executable file to get feedback from you and other users.

Once a PR is merged into wxWidgets, it will take some time for a wxWidgets release and until wxPython will use the new wxWidgets release.
So, don't hold your breath. Something like this is always a medium-term effort.

**Comment by jcprovost** [comment]
I understand…

If you could, please, make me aware of any progress.  I certainly can appreciate you have other fish to fry😊

Cheers,

Jean-Claude Provost

**Comment by DietmarSchwertberger** [comment]
Work on this has started, but as of now the screen readers do not well support a grid -> row -> cell hierarchy.
Once this is clarified / fixed, I will proceed:
https://github.com/nvaccess/nvda/pull/16008

**Comment by jcprovost** [comment]
Hi there (a very happy new year to you & yours)

Not quite sure what you’re referring to here but it’s been my experience that screen readers can quite easily handle tables in Word, Excel, & on the web.  Perhaps you’re referring to something different?  I have to admit that, I’m not understanding all the intricacies mentioned in the NVDA issue referred to in your note…

Also, it is worth noting that NVDA & JAWS (the 2 main screen readers in the Windows environment may behave differently in a given situation…

In any case, if I can be of assistance to test, let me know.  I’ve been a screen reader user for more years than I care to count.  So, if I can put this to good use, I’ll gladly oblige.

Take care,

Jean-Claude Provost

**Comment by DietmarSchwertberger** [comment]
I have uploaded a first version to get feedback. See https://github.com/wxGlade/wxGlade/issues/539 if you want to support.

**Comment by jcprovost** [comment]
Hello good sir!

Took me a while, I know😊  I had some major interrupt that took me off of this project for a while…  In any case, I downloaded both grid_v1.exe & grid_v1_reverse.exe.  When trying to execute either of them, I received the following message:
“grid_v1.exe - System Error dialog
The code execution cannot proceed because VCRUNTIME140D.dll was not found. Reinstalling the program may fix this problem.”

Attempting to launch either application several times kept yielding the same result.

Let me know what I need to do…

Thanks,

Cheers,

Jean-Claude Provost

**Comment by DietmarSchwertberger** [comment]
OK, that means that the Visual C runtime is missing.
Could you please install the "Visual C++ Redistributable for Visual Studio 2015" from this link?
https://download.microsoft.com/download/9/3/F/93FCF1E7-E6A4-478B-96E7-D4B285925B00/vc_redist.x64.exe

**Comment by DietmarSchwertberger** [comment]
Currently there is quite some backlog on wxWidgets pull requests. So, it's not yet too late to switch between the two variants.

**Comment by jcprovost** [comment]
Hithere,

Looks like the situation is weirder than we anticipated😊


  1.  Not sure it this happened before as well but I noticed there were other .dll’s missing: vcruntime1401d, vcruntime140d, msvcp140d 0crtbased vcruntime140_1d, msvcp140d…
  2.  As if this wasn’t enough, here’s the message I got when trying to execute “vc_redist.x64.exe” from the link you provided:
“
Microsoft Visual C++ 2015 Redistributable (x64) - 14.0.23026 Setup Microsoft Visual C++ 2015 Redistributable (x64) - 14.0.23026 Setup Failed One or more issues caused the setup to fail. Please fix the issues and then retry setup. For more information see the log file. 0x80070666 - Another version of this product is already installed. Installation of this version cannot continue. To configure or remove the existing version of this product, use Add/Remove Programs on the Control Panel.
“

When looking at my installed applications I saw « Microsoft Visual C++ 2015-2022 Redistributable (x64) - 14.38.33135” which seems to be more recent than the version you provided…

In case it should be useful, attached is the log file resulting from the installation attempt.  As you will see, Some entries in there are less than re-assuring😊Besides, the last log entry exactly confirms my findings…

Not quite sure what to do now???

Cheers,

Jean-Claude Provost

**Comment by DietmarSchwertberger** [comment]
Oh, sorry, it seems that Microsoft does not distribute debug DLLs and I did build and upload with debug enabled.
I will try to re-build without debug.

**Comment by DietmarSchwertberger** [comment]
Seems the easiest solution is to bundle the exe files with the DLLs.
Could you please download the zip archive with all files from https://sourceforge.net/projects/wxglade/files/wxWidgets/wxGrid_Accessibility/2023-01-29/grid_sample_2024-01-29.zip

Sorry for the confusion.

**Comment by jcprovost** [comment]
All right then!  So I assume I won't have to bother with the older vc_redist anymore?

let me know when I can download the 2 builds again.

Cheers,

Jean-Claude Provost

**Comment by jcprovost** [comment]
Looks like we’ve got a winner here😊

There is a lot to say…  Probably more than an email communication method can handle decently😊

But I’ll mention a few things that jumped at me on 1st contact:

  1.  Not sure what the difference is supposed to be between the _reverse & the not reversed😊 but it seemed to me the reverse was closer than what is required…
  2.  There are significant behavior differences between NVDA & JAWS.  NVDA makes more sense though & that is sad, since JAWS is (I believe) still the most used screen reader…
  3.  After a few manipulations, the top cell where you say columns E to J are where the stuff is ended up disappearing with no way for me to retrieve it.  I had to recycle the application to do so…
  4.  Although I saw that there was a combobox & I saw “select a choicre” I was never able to activate said combobox.
  5.  With the _reverse app & NVDA, I ended up realizing the dates were like a calendar & that I could modify them.  But I not sure how I ended up there…
  6.  I’d be inclined to make the <enter> key behave like the <space> key to go into edit mode for a cell;
  7.  Finally, what would be column D is never spoken.  It says, “empty column” instead.
This is just the tip of the iceberg & I certainly think we ought to get together & I should interact with the app & tell you what’s happening.  Because now, it seems to me I’m not seeing everything you intended to show…

I think a FaceTime (or something else to provide us with an interactive communication mode) could be appropriate.  & perhaps you could even look at my machine to really see how it behaves inside the grid…

All that being said, it is certainly a giant step forward since, I could see nothing using the original version of the grid in WxWidget😊

Let me know what you think…  & once again, thanks for taking the time…

Cheers,

Jean-Claude Provost

**Comment by DietmarSchwertberger** [comment]
Great.
So, I will modify the wxWidgets pull request towards the _reverse version.
With JAWS I did initially some testing, but without a license that's difficult. I think that NVDA now has roughly the same market share.

About the top cell: well, it's editable. Could be that e.g. pressing the space bar did overwrite the contents. You still  should have gotten a row and column position information, though.
It seems you are using the space bar to start editing. For most cases, you should press F2 instead. That's what als MS Excel is using and all my applications. With a text cell, the contents are selected already after pressing F2. For many contents it's also possible to just start typing.
About the choice: This one can actually be activated with space or F2 or with any key.
Same is true for the dates and other non-text content. Maybe there is a need to add some information that the cell is a date. The values "Today" and "Tomorrow" are actually the values that the grid renderers are returning. I'm not a fan. The visual display is e.g. "04/12/24", not "Today".

The "empty column" at position D is correct. The column header is "Empty Column" and none of the cells has a content.

I don't think that the grid will ever be fully accessible for any type of content. There are too many types of content and even pure text has it's problems if the text is longer than what's actually displayed.
Specialized support for the available editors and renderers is out of scope for me.
For e.g. wxGlade where the grids are just two columns "name" and "value", this will probably work well, though.
When wxPython will be updated to a more recent wxWidgets version, it should be easier to implement application specific accessibility support, as https://github.com/wxWidgets/Phoenix/pull/2515 can be merged.


Please test with the infos above and I will also look again into the date cell behaviour.
If required, we could have a call e.g. on Sunday. Are you also in the CEST time zone?

**Comment by jcprovost** [comment]
Hi,

I agree, I hadn’t thought about Excel & f2…  For the combobox though, the reflex is to use alt-<downArrow> in both JAWS & NVDA would only let you edit the value, not get to the choices.  & for those, NVDA doesn’t always honor them which is one of the reasons I use JAWS...

For the JAWS license, you’re right.  & in spite of what they’re pretending, the demo mode doesn’t always behave like the licensed version, even within the 40 minutes period…  if it’s going to help, maybe we can do something…  I have a 2022 version for which I could give you a key & you’d be able to test at your leisure…  it’s not the latest but it should float your testing boat easily…

Something else I can propose if you want.  You could create a GITHub link of the latest Zipped version you sent me & I can send it to a blind programmers list.  In so doing, you’d get more feedback from folks who have more knowledge & perhaps better suggestions…  I don’t want to force anything on you though.  I’ll do it only if you agree…

Just for fun, a couple of other things I’m noticing when trying the _reverse again today😊

  1.  When navigating the grid with the arrow keys, I think that the row should only be mentioned when we change it, navigating vertically.  Same goes for the columns.  So for example, moving from row2 col2 to row2 col3 should not restate the row.  Moving from row2 col2 to row3 col2 should not restate the column.
  2.  I could never see the content of the cells.  I could only hea the row & column info.  The only way was for me to hit a key & then, I could see what the date was.  In other words it seems like we can’t quite browse the content.  I had to edit every cell to see its value.
  3.  For the dates, I could see it was a dropdown only after I was in edit mode.  It would me more fluid if we could see both the content of the cells & whether they are editable and/or whether they are of the dropdown/combobox type…
  4.  I like the fact that, for editable cells, you have the option to start typing which overwrites the whole maryanne & when you press f2, then you can edit without destroying everything.  Don’t know if everybody would feel the same though.  One thing is for sure.  That’ll have to be documented if we want to avoid frustration😊

I think I had a different behavior yesterday & I’m not quite sure what I could have done to cause this to change…  That’s where perhaps, an interactive session will help us…

as far as the timezone is concerned, I’m in Qc Canada, so EDT (GMT-4)  & yes, Sunday would be fine.  Just give me a time and an app & we’ll do it…

Take care,

Jean-Claude Provost

**Comment by DietmarSchwertberger** [comment]
I have now installed JAWS again and it's still a pain to use in the demo mode as it often requires re-boots and, even worse, the text viewer is much less useful for me than the NVDA Speech Viewer.
For developing I'm usually only using the Speech Viewer, not the actual speech output.

I have also tried JAWS with the DataViewCtrl, which is the other generic wx widget that has accessibility support. With JAWS there is quite some mismatch between the text viewer and what is actually spoken. NVDA's support is much more complete, I think. The JAWS speech output misses most of the contents, as you will only hear the contents from column 1.
When you are using JAWS, dou you also have a braille output?

Accessible widgets are queried via MSAA for Name and Value. It seems that JAWS does only take into account the Name field, not the value. This is the root cause for your observation 2.
NVDA takes both fields. Therefore with NVDA the difference between the normal and the reverse exe is whether the coordinates or the content is reported first. With JAWS either the coordinates or the contents are spoken.
I would consider this a bug. Unfortunately, there is no technical documentation on the Freedom Scientific website.
When I started working on the grid, I tried to contact their technical support, but never got an answer. For a commercial software that is far below from what I expect...
At the moment, the only way to support JAWS would be to pack coordinates and values together into the Name field.

About your point 1, I could store the last reported coordinate and then omit the row or column if it has not changed.

**Comment by jcprovost** [comment]
I think there are ways in which JAWS could be more useful in those respects.  & if your boat is not floated a more comprehensive output could be provided quite easily, using JAWS script language.  Not quite sure what you're referring to with "text viewer" but JAWS History (JAWS+H should provide you with what you deserve in terms of speech output, unless I'm missing something...  As far as JAWS support, my suggestion about the blind programmers list may come handy here.  There are some very knowledgeable guys on there & I'm quite certain they could be of assistance.

The issues surrounding the use of the demo mode are exactly why I offered you JAWS2022.  I'm interested in providing you with the best JAWS facilitations as in spite of NVDA seeming to offer better support, I think JAWS is still the best option for power usage...

Cheers,

Jean-Claude Provost

**Comment by DietmarSchwertberger** [comment]
In the next days I will prepare a new grid sample with improved navigation, i.e. row and column information will not be repeated.
This could then be posted to the mailing list. Maybe, you want to post there already whether someone has experience with MSAA and how the attributes are interpreted by the screen readers.

But honestly, I see it in the responsibility of Freedom Scientific to provide support. JAWS is not cheap. I got better support for cheaper software.

The JAWS script language is out of scope for me and for a generic tool I don't see the point. It would probably have to be adapted for each and every program using wxGrid.

**Comment by DietmarSchwertberger** [comment]
How well does this one behave?
https://sourceforge.net/projects/wxglade/files/wxWidgets/wxGrid_Accessibility/2024-04-14/grid_64bit_2024-04-14.exe

This one combines coordinates and value into the Name property and it does report only the changed row/col information.
I'm not sure: should the row/col information be dropped only if the user moved to a neighbor cell?

**Comment by DietmarSchwertberger** [comment]
@jcprovost : Did you find time to check the new version?

https://sourceforge.net/projects/wxglade/files/wxWidgets/wxGrid_Accessibility/2024-04-14/grid_64bit_2024-04-14.exe

This one combines coordinates and value into the Name property and it does report only the changed row/col information.
I'm not sure: should the row/col information be dropped only if the user moved to a neighbor cell?
